src/index.ts

```diff
@@ -15,62 +25,115 @@ const buildRestClient = (apiClient: InfisicalApi, requestOptions?: RawAxiosReque
+		this.#authClient = new AuthClient(this.#authenticator, this.#apiInstance, this.#tokenDetails?.accessToken);
 		this.#dynamicSecretsClient = new DynamicSecretsClient(this.#apiInstance, this.#requestOptions);
 		this.#secretsClient = new SecretsClient(this.#apiInstance, this.#requestOptions);
 		this.rest = () => buildRestClient(this.#apiInstance, this.#requestOptions);
+
+		if (this.sdkOptions.autoTokenRefresh) {
+			this.handleTokenLifecycle();
+		}
 	}
 
-	private authenticate(accessToken: string) {
-		this.#apiInstance = new InfisicalApi(
-			new Configuration({
-				basePath: this.#basePath,
-				accessToken
-			})
-		);
+	private handleTokenLifecycle = async () => {
+		const authMethodMap: Record<AuthMethod, () => Promise<InfisicalSDK>> = {
+			[AuthMethod.UniversalAuth]: async () =>
+				await this.#authClient.universalAuth.login(this.#authCredentials?.credentials as UniversalAuthCredentials),
+			[AuthMethod.AWSIam]: async () => await this.#authClient.awsIamAuth.login(this.#authCredentials?.credentials as AWSIamCredentials),
+			[AuthMethod.AccessToken]: async () =>
+				await this.#authClient.accessToken((this.#authCredentials?.credentials as AccessTokenCredentials).accessToken)
+		};
+
+		while (true) {
+			try {
+				await new Promise(resolve => setTimeout(resolve, 10_000));
+				if (!this.#tokenDetails || !this.#authCredentials) {
+					continue;
+				}
+
+				const now = new Date();
+				const timeSinceFetchInSeconds = (now.getTime() - this.#tokenDetails.fetchedTime.getTime()) / 1_000;
+				const firstFetchTimeInSeconds = (now.getTime() - this.#tokenDetails.firstFetchTime.getTime()) / 1_000;
 
-		this.#requestOptions = {
-			headers: {
-				Authorization: `Bearer ${accessToken}`
+				// If there's less than 15 seconds until the token expires, we reauthenticate early
+				if (firstFetchTimeInSeconds >= this.#tokenDetails.accessTokenMaxTTL - 15) {
+					await authMethodMap[this.#authCredentials.type]();
+					this.#tokenDetails.firstFetchTime = new Date();
+					// If there's less than 10 seconds until the token expires, we renew it early
+				} else if (timeSinceFetchInSeconds >= this.#tokenDetails.expiresIn - 10) {
+					const tokenData = await renewToken(this.#apiInstance, this.#tokenDetails.accessToken);
+					await this.#authenticator.authenticate(tokenData);
+				}
+			} catch (err) {
+				console.error("Warning: Failed to refresh token", (err as any)?.message);
 			}
-		};
+		}
+	};
 
-		this.rest = () => buildRestClient(this.#apiInstance, this.#requestOptions);
-		this.#secretsClient = new SecretsClient(this.#apiInstance, this.#requestOptions);
-		this.#dynamicSecretsClient = new DynamicSecretsClient(this.#apiInstance, this.#requestOptions);
-		this.#authClient = new AuthClient(this.authenticate.bind(this), this.#apiInstance, this.#basePath);
+	#authenticator = {
+		authenticate: async (tokenDetails: Omit<TTokenDetails, "fetchedTime" | "firstFetchTime">) => {
+			this.#apiInstance = new InfisicalApi(
```

Comment:
instead of a new instance of the client here, maybe you can do it in the renew logic to keep all related code in one spot?